delphi/translators/for2py/rectify.py

```diff
@@ -0,0 +1,1982 @@
+"""
+The purpose of this program is to do all the clean up for translate.py.
+This (rectify.py) program will receive OFP generated XML file as an input.
+Then, it removes any unnecessary elements and refactor randomly structured
+(nested) elementss into a correct structure. The output file will be
+approximately 30%~40% lighter in terms of number of lines than the OFP XML.
+
+Example:
+    This script is executed by the autoTranslate script as one
+    of the steps in converted a Fortran source file to Python
+    file. For standalone execution:::
+
+        $python rectify.py <ast_file>
+
+ast_file: The XML represenatation of the AST of the Fortran file. This is
+produced by the OpenFortranParser.
+
+Author: Terrence J. Lim
+Last Modified: 4/11/2019
+"""
+
+import sys
+import re
+import argparse
+import xml.etree.ElementTree as ET
+
+
+class RectifyOFPXML:
+    def __init__(self):
+        self.is_derived_type = False
+        self.is_derived_type_ref = False
+        self.is_array = False
+        self.is_format = False
+        self.is_function_arg = False
+        self.need_reconstruct = False
+
+        self.cur_derived_type_name = ""
+        self.current_scope = ""
+
+        # Keep a track of both array and non-array variables in the dictionary
+        # of {'name' : 'scope'}
+        self.declared_non_array_vars = {}
+        self.declared_array_vars = {}
+
+        self.derived_type_var_holder_list = []
+        self.derived_type_refs = []
+        self.subscripts_holder = []
+
+        self.format_holder = ET.Element("")
+        self.parent_type = ET.Element("")
+        self.derived_type_ref = ET.Element("")
+        self.current_body_scope = ET.Element("")
+
+    """
+        Nested child tag list
+    """
+    FILE_CHILD_TAGS = ["program", "subroutine", "module"]
+
+    STATEMENT_CHILD_TAGS = [
+        "assignment",
+        "write",
+        "format",
+        "stop",
+        "execution-part",
+        "print",
+        "open",
+        "read",
+        "close",
+        "call",
+        "statement",
+        "label",
+        "literal",
+        "continue-stmt",
+        "do-term-action-stmt",
+        "return",
+        "contains-stmt",
+        "declaration",
+        "prefix",
+        "function",
+        "internal-subprogram",
+        "internal-subprogram-part",
+        "prefix",
+        "exit",
+    ]
+
+    LOOP_CHILD_TAGS = ["header", "body", "format"]
+
+    DECLARATION_CHILD_TAGS = [
```

Comment:
@hlim1 Usually, the convention is to put global variables in all uppercase letters - are these meant to be global variables (i.e. are they never going to change during the execution of the program?) If so, can you move them outside the class definition (see pyTranslate.py for examples).